packages/react-query/src/internals/getQueryKey.ts

```diff
@@ -1,11 +1,100 @@
+import {
+  AnyMutationProcedure,
+  AnyQueryProcedure,
+  AnyRouter,
+  DeepPartial,
+  inferProcedureInput,
+} from '@trpc/server';
+import { QueryType, getArrayQueryKey } from '../internals/getArrayQueryKey';
+import { DecorateProcedure, DecoratedProcedureRecord } from '../shared';
+
 /**
  * We treat `undefined` as an input the same as omitting an `input`
  * https://github.com/trpc/trpc/issues/2290
  */
-export function getQueryKey(
+export function getQueryKeyInternal(
   path: string,
   input: unknown,
 ): [string] | [string, unknown] {
   if (path.length) return input === undefined ? [path] : [path, input];
   return [] as unknown as [string];
 }
+
+type GetInfinitQueryInput<TProcedureOrRouter extends AnyQueryProcedure> =
+  Record<never, never> extends Omit<
+    inferProcedureInput<TProcedureOrRouter>,
+    'cursor'
+  >
+    ? undefined
+    :
+        | DeepPartial<Omit<inferProcedureInput<TProcedureOrRouter>, 'cursor'>>
+        | undefined;
```

Comment:
Here are two tricks that may improve perf and readability
```suggestion
type GetInfinitQueryInput<TProcedureOrRouter extends AnyQueryProcedure, 
TInputWithoutCursor = Omit<
    inferProcedureInput<TProcedureOrRouter>,
    'cursor'
  >> =  keyof TInputWithoutCursor extends never
    ? undefined
    : DeepPartial<TInputWithoutCursor> | undefined;
```

Reply:
TInput might not be an object so Omit without check wont work - right?

Reply:
The check is made above, so it's guarantied to work: 
```ts
type GetQueryProcedureInput<TProcedureOrRouter extends AnyQueryProcedure> =
  inferProcedureInput<TProcedureOrRouter> extends {
    cursor?: any;
  }
    ? GetInfinitQueryInput<TProcedureOrRouter>
    : DeepPartial<inferProcedureInput<TProcedureOrRouter>> | undefined;
```

Reply:
i can never get on a laptop on weekends but will take a look tomorrow! appreciate you @ecyrbe